software/bootcamp-code/main.py

```diff
@@ -33,56 +52,183 @@ def main_loop():
-            # objects of interest and calculate the required motion for reaching the object.
-
-            # STATE TO FIND THE BALL 0
+            
+            # Here is the state machine for the driving logic.
+                #Omni-motion movement:
+                # side speed is x_speed
+                # forward speed is y_speed
+                # rotation if you want to turn
+            
+            # STATE TO FIND THE BALL
             if state == State.FIND_BALL:
+                # if we have a ball in view, center it
                 if len(processedData.balls)>0:
-                    state=State.CENTER_BALL
+                    state=State.MOVE_CENTER_BALL
+                    continue
+                # if no ball found, rotate
+                
+                omni_motion.move(0, 0, find_rotation_speed)
+
+            # STATE TO MOVE TO AND CENTER THE BALL
+            elif state == State.MOVE_CENTER_BALL:
+                # if there are no balls, then find one
+                if len(processedData.balls)<1:
+                    state=State.FIND_BALL
                     continue
-                rotation = 15
-                omni_motion.move(0, 0, rotation)
-
-            # STATE TO DRIVE TO BALL is 1
-            elif state == State.CENTER_BALL:
-                # The biggest ball is the first one in the detected balls list.
-                # Find the coordinates of the largest ball.
-                # (THE ZERO COORDINATES OF THE FRAME ARE IN THE UPPER LEFT CORNER.)
-                # Using these coordinates calculate the side speed, forward speed and rotation for the robot.
+                    
+                # The biggest ball is the last one in the detected balls list.
+                # Using the coordinates of the biggest ball calculate the side speed, forward speed and rotation for the robot.
                 
                 # The destination coordinates are the difference between the ball location and desired location.
-                dest_x = ball_desired_x - processedData.balls[0].x
-                dest_y = ball_desired_y - processedData.balls[0].y
-                # Normalize destination in the [-1;1] range and and then multiply it with motor max speed.
-                # (then the driving is proportional)
-                # Speed range for motors is 48 - 2047, we use 100 for max motor speed at the moment.
-                x_speed = (dest_x/cam.rgb_width) * 100
-                y_speed = (dest_y/cam.rgb_height) * 100
-                #rotation = 0
-                # Drive towards a ball:
-                    # side speed is x_speed
-                    # forward speed is y_speed
-                    # rotation if you want to turn
-                #print("X: ", x_speed, "Y: ", y_speed)
-                omni_motion.move(x_speed, y_speed, 0)
+                dest_x = ball_desired_x - processedData.balls[-1].x
```

Comment:
I also suggest writing a method that normalizes the ball x or y locations so you have to write less duplicate code.

Reply:
Added.